File: Advance/18.Implement Kruskal’s Algorithm for finding the Minimum Spanning Tree (MST)/Solution.java
import java.util.*;

class KruskalAlgorithm {
    static class Edge implements Comparable<Edge> {
        int u, v, weight;

        Edge(int u, int v, int weight) {
            this.u = u;
            this.v = v;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge other) {
            return this.weight - other.weight;
        }
    }

    static class UnionFind {
        int[] parent, rank;

        UnionFind(int size) {
            parent = new int[size];
            rank = new int[size];
            for (int i = 0; i < size; i++) {
                parent[i] = i;
                rank[i] = 0;
            }
        }

        int find(int x) {
            if (parent[x] != x)
                parent[x] = find(parent[x]);
            return parent[x];
        }

        void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);

            if (rootX != rootY) {
                if (rank[rootX] < rank[rootY]) {
                    parent[rootX] = rootY;
                } else if (rank[rootX] > rank[rootY]) {
                    parent[rootY] = rootX;
                } else {
                    parent[rootY] = rootX;
                    rank[rootX]++;
                }
            }
        }
    }

    static void kruskal(List<Edge> edges, int V) {
        Collections.sort(edges);

        UnionFind uf = new UnionFind(V);
        int mstWeight = 0;
        System.out.println("Edges in the MST:");

        for (Edge edge : edges) {
            int u = edge.u;
            int v = edge.v;

            if (uf.find(u) != uf.find(v)) {
                System.out.println(u + " -- " + v + " == " + edge.weight);
                mstWeight += edge.weight;
                uf.union(u, v);
            }
        }

        System.out.println("Total weight of MST: " + mstWeight);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the number of vertices and edges: ");
        int V = scanner.nextInt();
        int E = scanner.nextInt();

        List<Edge> edges = new ArrayList<>();
        System.out.println("Enter the edges (u, v, weight):");
        for (int i = 0; i < E; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            int weight = scanner.nextInt();
            edges.add(new Edge(u, v, weight));
        }

        kruskal(edges, V);
    }
}

File: Advance/18.Implement Kruskal’s Algorithm for finding the Minimum Spanning Tree (MST)/Solution.py
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX == rootY:
            return
        if self.rank[rootX] < self.rank[rootY]:
            self.parent[rootX] = rootY
        elif self.rank[rootX] > self.rank[rootY]:
            self.parent[rootY] = rootX
        else:
            self.parent[rootY] = rootX
            self.rank[rootX] += 1

def kruskal(V, edges):
    # Sort edges based on weight
    edges.sort(key=lambda edge: edge[2])
    ds = DisjointSet(V)
    mst = []
    total_weight = 0

    for u, v, w in edges:
        if ds.find(u) != ds.find(v):
            ds.union(u, v)
            mst.append((u, v, w))
            total_weight += w

    return mst, total_weight

if __name__ == "__main__":
    V = int(input("Enter the number of vertices: "))
    E = int(input("Enter the number of edges: "))
    
    edges = []
    print("Enter each edge in the format: u v weight (vertices are 0-indexed):")
    for _ in range(E):
        u, v, w = map(int, input().split())
        edges.append((u, v, w))
    
    mst, total_weight = kruskal(V, edges)
    
    print("\nEdges in the Minimum Spanning Tree (MST):")
    for u, v, w in mst:
        print(f"{u} -- {v} == {w}")
    print("Total weight of MST:", total_weight)
